# Agenda:
1. Python Itertools
2. Python Collections
3. Some Handy One-liners

## 1. Python Itertools

- Set of functions to build effective iterators in Python. Combines them to form an "iterator algebra".
- Iterators & Iterables?

**Iterators** and **Iterables** have a clear difference in meaning when it comes to Python. In very simple terms, **iterables** are objects on which you can iterate upon. For example, list, tuple, and dictionary are Python’s built-in iterable container data types. 

So, when we want to iterate upon these iterables, we write a “for loop” or a list/dictionary comprehension. And with these, Python internally creates what we call as Iterators which essentially are responsible for “Iteration”. 

Strictly speaking, Iterables are objects that contain the `__iter__` method, which return an iterator. Now, to iterate upon the iterable, the iterators have the `__next__` method which is responsible to fetch the next item from the iterable and track the indexes.

## What is the Itertools module?

“Iter” in Itertools stands for Iterables. The itertools module in Python offers a set of tools or functions which are designed for a specific task but can also be used in combinations to build a more complex iterator. 

You can think of these as blocks of efficient functions which make our lives easy as we don’t have to write functions for common and repetitive tasks manually.  And you can also use them as an “Iterator Algebra” when you combine some of these together for your use cases.

So with these functions, you can create clean, efficient and smart code rather than creating a messy and less efficient one.


In [ ]:
import itertools

1. zip and zip_longest

`zip_longest()` is a function of the itertools module which lets you work with multiple iterables of different sizes while zipping. With zip, the iterable of the shortest length is considered and the rest of the elements of all other iterables are ignored. Let’s see how we can tackle this with zip_longest.

In [3]:
a = [1,2,3]
b = [3,5,6,7,8]
lis = []
for i,j in zip(a,b):
  lis.append((i,j))

In [4]:
lis

[(1, 3), (2, 5), (3, 6)]

You see? It ignored values 7 and 8.

Now let’s see how zip_longest works.


In [5]:
from itertools import  zip_longest

In [6]:
a = [2,3,4,5]
b = [3,5,6,7,8,9,2]
lis = []
for i,j in zip_longest(a,b,fillvalue="X"):
  lis.append((i,j))
print(lis)

[(2, 3), (3, 5), (4, 6), (5, 7), ('X', 8), ('X', 9), ('X', 2)]


So the argument `fillvalue` is needed to be passed to specify with which value you need Python to fill the shorter iterables with.

#### Infinite iterators

2. Count

In [8]:
import itertools

In [9]:
counter = itertools.count()

In [ ]:
print(next(counter))

In [ ]:
for i in itertools.count(1,2):
    print(i)

In [11]:
counter2 = itertools.count(3, -1)

In [14]:
print(next(counter2))

1


3. Cycle

In [15]:
cycler = itertools.cycle('ABCD')

In [20]:
print(next(cycler))

A


In [21]:
for i in cycler:
  print(i)

Streaming output truncated to the last 5000 lines.
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A
B
C
D
A

KeyboardInterrupt: ignored

4. Repeat

In [22]:
repeater = itertools.repeat(10, 4)

In [27]:
print(next(repeater))

StopIteration: ignored

In [28]:
sqr = map(pow, range(4), itertools.repeat(2,4))
sqr

In [32]:
next(sqr)

9

In [33]:
repeater2 = itertools.repeat(4,5)

In [34]:
for i in repeater2:
  print(i)

4
4
4
4
4


5. Accumulate

The `accumulate()` function takes in an iterable and a function which does some operation on the elements of the iterables. The results are then **accumulated** and operated upon. Sounds confusing? Let’s have a look at the code

In [35]:
from operator import mul

In [37]:
list_a = [1, 2, 3, 4, 5]

In [38]:
list(itertools.accumulate(list_a, mul))

[1, 2, 6, 24, 120]

So accumulate took the list A and a function from the operator module which multiplies two numbers. Then it iterates and produces output in following manner:
1. 1 is the first element, so it as is 1
2. 1 mul by 2 is 2
3. 2 mul by 3 is 6
4. 6 mul by 4 is 24
5. 24 mul by 5 is 120


lambda

In [41]:
#using lambda
list(itertools.accumulate(list_a))

[1, 3, 6, 10, 15]

6. Chain

In [43]:
fruits = ['mango', 'apple', 'banana'] #10M
prices = [10, 20, 8]
quantity = [3, 5, 9]

In [44]:
chained = itertools.chain(fruits, prices, quantity)

In [ ]:
next(chained)

In [45]:
list(itertools.chain(fruits, prices, quantity))

['mango', 'apple', 'banana', 10, 20, 8, 3, 5, 9]

Memory efficient

In [46]:
list(itertools.chain(list_a, 'ABCSD'))

[1, 2, 3, 4, 5, 'A', 'B', 'C', 'S', 'D']

7. Filterfalse

The filterfalse function does exactly what its name is- it filters the elements which give false against a certain condition. That is, only the elements not matching to the condition will be in the output. 

In [47]:
list(itertools.filterfalse(lambda x: x%2==0, range(10)))

[1, 3, 5, 7, 9]

8. Starmap

**Starmap** function creates an iterator that takes in a function and an iterable with iterables within it as arguments. The function is then applied on the iterables inside just like the map function.
The difference between map and starmap is, literally, of a “star”. What does this star do?
Answer: **Unpacking**.


In [48]:
lis = [(2, 2), (3, 2), (4, 2)]
print(list(map(pow, lis)))

TypeError: ignored

In [49]:
print(list(itertools.starmap(pow, lis)))

[4, 9, 16]


### Combinatorial Iterators

In [ ]:
list(itertools.product('AB', repeat=2))

[('A', 'A'), ('A', 'B'), ('B', 'A'), ('B', 'B')]

In [ ]:
list(itertools.combinations('ABC', 2))

[('A', 'B'), ('A', 'C'), ('B', 'C')]

In [ ]:
list(itertools.combinations_with_replacement('ABC', 2))

[('A', 'A'), ('A', 'B'), ('A', 'C'), ('B', 'B'), ('B', 'C'), ('C', 'C')]

In [ ]:
list(itertools.permutations('ABC', 2))

[('A', 'B'), ('A', 'C'), ('B', 'A'), ('B', 'C'), ('C', 'A'), ('C', 'B')]

# 2. Python Collections

The Collections module in Python consists of special container data types which extend the features and abilities of stock containers like lists, tuples and dictionaries.

With these container data types, you not only have features of these stock containers, but also some specialized features. Let's go over these one by one.

In [50]:
import collections

1. Counter

It is a collection where elements are stored as dictionary keys and their counts are stored as dictionary values.

In [51]:
Marvel = "Bad Wolverine bullied Iron Man Bad Bad Wolverine Poor Poor Iron Man"

In [52]:
Marvel_count = collections.Counter(Marvel.split())

In [53]:
type(Marvel_count)

collections.Counter

In [55]:
Marvel_count['Bad']

3

In [56]:
Marvel_count.elements

<bound method Counter.elements of Counter({'Bad': 3, 'Wolverine': 2, 'Iron': 2, 'Man': 2, 'Poor': 2, 'bullied': 1})>

In [57]:
Marvel_count.values()

dict_values([3, 2, 1, 2, 2, 2])

In [58]:
Marvel_count.keys()

dict_keys(['Bad', 'Wolverine', 'bullied', 'Iron', 'Man', 'Poor'])

In [60]:
Marvel_count['Wolverine']

2

In [63]:
Marvel_count.most_common(5)

[('Bad', 3), ('Wolverine', 2), ('Iron', 2), ('Man', 2), ('Poor', 2)]

In [64]:
Marvel_2 = collections.Counter({'Bad': 1, 'Wolverine': 2, 'Iron': 2, 'Man': 2, 'Poor': 2, 'bullied': 1})

In [65]:
Marvel_count.subtract(Marvel_2)

In [ ]:
Marvel_count.update(Marvel_2)

In [66]:
Marvel_count

Counter({'Bad': 2,
         'Iron': 0,
         'Man': 0,
         'Poor': 0,
         'Wolverine': 0,
         'bullied': 0})

2. ChainMap

A ChainMap groups multiple dicts or other mappings together to create a single, updateable view. If no maps are specified, a single empty dictionary is provided so that a new chain always has at least one mapping.

The underlying mappings are stored in a list. That list is public and can be accessed or updated using the maps attribute. There is no other state.

Lookups search the underlying mappings successively until a key is found. In contrast, writes, updates, and deletions only operate on the first mapping.

A ChainMap incorporates the underlying mappings by reference. So, if one of the underlying mappings gets updated, those changes will be reflected in ChainMap.

All of the usual dictionary methods are supported. In addition, there is a maps attribute, a method for creating new subcontexts, and a property for accessing all but the first mapping:

In [68]:
dic1 = { 'a' : 1, 'b' : 2 } 
dic2 = { 'b' : 3, 'c' : 4 } 
dic3 = { 'd' : 5, 'b' : 7 }

In [69]:
chain1 = collections.ChainMap(dic2, dic1)

In [71]:
chain1
# dic2 --> dic1

ChainMap({'b': 3, 'c': 4}, {'a': 1, 'b': 2})

In [70]:
list(chain1)

['a', 'c', 'b']

In [76]:
chain1.maps

[{'b': 3, 'c': 4}, {'a': 1, 'b': 2}]

In [75]:
list(chain1.keys())

['a', 'c', 'b']

In [ ]:
list(chain1.values())

[2, 1, 4]

In [ ]:
chain1 = chain1.new_child(dic3)

In [77]:
chain1.maps

[{'b': 3, 'c': 4}, {'a': 1, 'b': 2}]

In [ ]:
list(chain1)

['a', 'c', 'b', 'd']

In [ ]:
list(chain1.keys())

['a', 'c', 'b', 'd']

In [80]:
chain1['b']

2

In [79]:
chain1.maps = reversed(chain1.maps) 

In [ ]:
chain1.maps

In [ ]:
chain1['b']

2

In [ ]:
chain1.parents

ChainMap({'b': 3, 'c': 4}, {'d': 5, 'b': 7})

3. Deque

Deques are a generalization of stacks and queues (the name is pronounced “deck” and is short for “double-ended queue”). 

In [81]:
deq = collections.deque([1, 2, 3, 4, 5])

In [82]:
deq

deque([1, 2, 3, 4, 5])

In [83]:
deq.append(6)

In [ ]:
deq.append(7)

In [ ]:
deq.append(8)

In [ ]:
deq

In [84]:
deq.appendleft(8)

In [85]:
deq

deque([8, 1, 2, 3, 4, 5, 6])

In [ ]:
deq.count(2)

1

In [86]:
deq.extend([0,0,0])

In [87]:
deq

deque([8, 1, 2, 3, 4, 5, 6, 0, 0, 0])

In [88]:
deq.extendleft([0,0,0])

In [89]:
deq

deque([0, 0, 0, 8, 1, 2, 3, 4, 5, 6, 0, 0, 0])

In [101]:
deq.insert(2,7)

In [102]:
deq

deque([0, 0, 7, 6, 5, 4, 3, 2, 1, 8, 0, 0])

In [92]:
deq.pop()

0

In [93]:
deq.popleft()

0

In [94]:
deq.remove(3)

In [ ]:
deq.remove(5)

In [ ]:
deq

deque([2, 3, 4])

In [95]:
deq.reverse()

In [96]:
deq

deque([0, 0, 6, 5, 4, 3, 2, 1, 8, 0, 0])

In [97]:
deq.rotate(2)

In [98]:
deq

deque([0, 0, 0, 0, 6, 5, 4, 3, 2, 1, 8])

In [99]:
deq.rotate(-2)

In [100]:
deq

deque([0, 0, 6, 5, 4, 3, 2, 1, 8, 0, 0])

4. Named Tuple

Named tuples assign meaning to each position in a tuple and allow for more readable, self-documenting code. They can be used wherever regular tuples are used, and they add the ability to access fields by name instead of position index.

In [103]:
Performance = collections.namedtuple('Employee_Rating', ['Q1', 'Q2', 'Q3', 'Q4'])

In [104]:
rahul = Performance(3,4,3.5,4.5)
ankit = Performance(4,4.5,4,4.5)

In [105]:
print(ankit)
print(rahul)

Employee_Rating(Q1=4, Q2=4.5, Q3=4, Q4=4.5)
Employee_Rating(Q1=3, Q2=4, Q3=3.5, Q4=4.5)


In [106]:
ankit[2]

4

In [108]:
ankit.Q3 > rahul.Q3

True

In [ ]:
ankit.Q1

4

In [ ]:
rahul.Q4

4.5

In [109]:
Milkha = Performance._make([4, 5, 5, 4.5])

In [ ]:
Milkha.Q1

4

In [110]:
print(Milkha)

Employee_Rating(Q1=4, Q2=5, Q3=5, Q4=4.5)


In [111]:
Performance._asdict(rahul)

OrderedDict([('Q1', 3), ('Q2', 4), ('Q3', 3.5), ('Q4', 4.5)])

In [112]:
rahul._replace(Q1=2)

Employee_Rating(Q1=2, Q2=4, Q3=3.5, Q4=4.5)

In [ ]:
rahul.Q1

In [113]:
Performance._fields

('Q1', 'Q2', 'Q3', 'Q4')

5. Defaultdict

A defaultdict works exactly like a normal dict, but it is initialized with a function (“default factory”) that takes no arguments and provides the default value for a nonexistent key.

A defaultdict will never raise a KeyError. Any key that does not exist gets the value returned by the default factory.

In [ ]:
dict2 = collections.defaultdict(list, {'a':1, 'b':2})

In [ ]:
dict2['a']

1

In [ ]:
dict2['a']+=1

In [ ]:
dict2

defaultdict(list, {'a': 2, 'b': 2, 'c': []})

In [ ]:
dict2['d']

[]

In [ ]:
dict2

defaultdict(list, {'a': 2, 'b': 2, 'c': [], 'd': []})

In [ ]:
marks = collections.defaultdict(lambda: 40)

In [ ]:
marks['Ankit'] = 97
marks['Rahul'] = 94
marks['Vikram']

40

In [ ]:
marks.items()

dict_items([('Ankit', 97), ('Rahul', 94), ('Vikram', 40)])

In [ ]:
s = 'mississippi'
d = collections.defaultdict(int)
for k in s:
  d[k] += 1

## Some Handy One Liners

Swapping

In [ ]:
a = 2
b = 4
a,b = b,a
print(a,b)

4 2


Palindrome

In [ ]:
str1 = 'abcba'
print(str1 == str1[::-1])

True


Flatenning 2 D array

In [ ]:
lis1 =  [ [1,2], [2,3], [3,4], [4,5] ]

In [ ]:
ls2 =  []

for i in lis1:
  for j in i:
    ls2.append(j)
print(ls2)

[1, 2, 2, 3, 3, 4, 4, 5]


In [ ]:
lis2 = [i for j in lis1 for i in j]
print(lis2)

[1, 2, 2, 3, 3, 4, 4, 5]


In [ ]:
sum(lis1,[])

[1, 2, 2, 3, 3, 4, 4, 5]

In [ ]:
lis3 = [[1], [2], [3], [4]]
sum(lis3,[])

[1, 2, 3, 4]

Reduce Function

In [ ]:
from functools import reduce
a = [1,2,3,4]

def sum(i,j):
  return i+j
  
print(reduce(sum,a))

#What it does :
#Step1: 1+2 = 3 (This 3 goes as the first element in the next operation)
#Step2: 3+3 = 6 (This 6 goes as the first element in the next operation)
#Step3: 6+4 = 10

10


In [ ]:
reduce(lambda x, y: x+y, a)

10